src/main/java/org/omnifaces/cdi/jsparam/JSParamExtension.java

```diff
@@ -0,0 +1,191 @@
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.enterprise.inject.spi.InjectionTarget;
+import javax.enterprise.inject.spi.ProcessInjectionTarget;
+import javax.faces.component.UIComponent;
+import javax.faces.component.html.HtmlBody;
+import javax.faces.component.html.HtmlCommandScript;
+import javax.faces.component.html.HtmlForm;
+import javax.faces.component.html.HtmlInputHidden;
+import javax.faces.event.AbortProcessingException;
+import javax.faces.event.ActionEvent;
+import javax.faces.event.ActionListener;
+import org.apache.commons.collections4.IteratorUtils;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.omnifaces.cdi.JSParam;
+import org.omnifaces.util.Components;
+import org.omnifaces.util.Faces;
+
+/**
+ * CDI extension that implements injection of client-side evaluated JavaScript results. The implementation is stateless,
+ * completely relying on the model provided and stored in the JSF component tree.
+ * <p>
+ * When initialized, the extension first checks if the body of the current view root contains a form suitable for
+ * passing the values collected from the specified javascript expressions. If a suitable {@link HtmlForm} form does not
+ * exist, one is created together with a {@link HtmlCommandScript}. The command script is also connected to a action
+ * listener that actually handles the injection logic - being called as soon as the page is loaded.
+ * Once done, the bean instance holding the injection points is checked for all<code>&#64;</code>{@link JSParam}
+ * properties defined in the class. Each property receives the JavaScript and hidden input fields needed to send the
+ * result of the evaluated expressions back to the CDI managed bean.
+ * 
+ * @since 3.3
+ * @author Adam Waldenberg, Ejwa Software/Hosting
+ */
+public class JSParamExtension implements Extension {
+	private static final int GENERATED_ID_NUM_CHARACTERS = 8;
+	private static final String JSPARAM_FORM_NAME = JSParam.class.getName().toLowerCase().replace('.', '-');
+	private static final String ASSIGNMENT_TEMPLATE = "document.getElementById('%s:%s').value = %s;";
+
+	private HtmlBody getDocumentBody() {
+		for (UIComponent c : Faces.getViewRoot().getChildren()) {
+			if (c instanceof HtmlBody) {
+				return (HtmlBody) c;
+			}
+		}
+
+		throw new IllegalStateException(String.format("The docoument %s does not contain a body",
+		                                              Faces.getContext().getViewRoot()));
+	}
+
+	private <T> void initializePageForm(T instance, Field field) {
+		if (Components.findComponent(JSPARAM_FORM_NAME) == null) {
+			final HtmlForm form = new HtmlForm();
+			final HtmlCommandScript commandScript = new HtmlCommandScript();
+
+			commandScript.addActionListener(new JSParamActionListener());
+			commandScript.setName(RandomStringUtils.randomAlphabetic(GENERATED_ID_NUM_CHARACTERS));
```

Comment:
This should have a more meaningful name too. Maybe based on the classes simple name.

Reply:
Yes, the randomness was actually added originally to support multiple forms - but in the end only one was really needed.